pkg/sort.go

```diff
@@ -0,0 +1,148 @@
+/*
+ripley
+Copyright (C) 2021  loveholidays
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+package ripley
+
+import (
+	"bufio"
+	"container/heap"
+	"encoding/json"
+	"fmt"
+	"os"
+	"time"
+)
+
+type requestSort struct {
+	requestHeap *requestHeap
+	// Used to check if requests are being emitted out of order
+	lastTimestamp time.Time
+	exitCode      int
+	// Number of requests to keep in the heap - once reached, start popping
+	bufferlen int
+	// Whether to panic on bad input or out of order timestamps
+	strict bool
+}
+
+func NewRequestSort(bufferlen int, strict bool) *requestSort {
+	rs := &requestSort{}
+	rs.requestHeap = &requestHeap{}
+	heap.Init(rs.requestHeap)
+	rs.lastTimestamp, _ = time.Parse(time.RFC3339, "1987-11-08T18:55:00.9Z")
+	rs.exitCode = 0
+	rs.bufferlen = bufferlen
+	rs.strict = strict
+	return rs
+}
+
+// An implementation of https://pkg.go.dev/container/heap
+// Holds `heapLen` requests sorted by timestamp and prints them in ascending order
+type requestHeap []*request
+
+func (rs requestHeap) Less(i, j int) bool {
+	return rs[i].Timestamp.Before(rs[j].Timestamp)
+}
+
+func (rs requestHeap) Len() int {
+	return len(rs)
+}
+
+func (rs *requestHeap) Pop() any {
+	old := *rs
+	n := len(old)
+	x := old[n-1]
+	*rs = old[0 : n-1]
+	return x
+}
+
+func (rs *requestHeap) Push(req any) {
```

Comment:
Why is it `any`? `requestHeap` is clearly typed.

Reply:
To adhere to the heap interface https://pkg.go.dev/container/heap#Push

Reply:
Oh, I see, it's a container/heap